nunjucks.1
.TH "NUNJUCKS" "1" "June 2025" "" ""
.SH "NAME"
\fBnunjucks\fR - Template Rendering Engine
.SH "SYNOPSIS"
.P
\fBnunjucks\fR \[lB]\fB-h\fR|\fB--help\fR\[rB] \[lB]\fB-V\fR|\fB--version\fR\[rB] \[lB]\fB-c\fR|\fB--config\fR \fIconfig-file\fR\[rB] \[lB]\fB-C\fR|\fB--option\fR \fIkey\fR=\fIvalue\fR\[rB] \[lB]\fB-d\fR|\fB--defines\fR \fIcontext-file\fR\[rB] \[lB]\fB-D\fR|\fB--define\fR \fIkey\fR=\fIvalue\fR\[rB] \[lB]\fB-e\fR|\fB--extension\fR \fImodule-name\fR\[rB] \[lB]\fB-o\fR|\fB--output\fR \fIoutput-file\fR|\fB-\fR\[rB] \[lB]\fIinput-file\fR|\fB-\fR\[rB]
.SH "DESCRIPTION"
.P
\fBnunjucks\fR(1) is a small command-line utility to render templates with the rich and powerful templating language \fBMozilla Nunjucks\fR \fI\(lahttps://mozilla.github.io/nunjucks/\(ra\fR. This allows you to define your configuration in a YAML file and then render an output file based on a template input file where your configuration can be expanded. It optionally can load Nunjucks addons like the ones from the companion \fBNunjucks Addons\fR \fI\(lahttps://github.com/rse/nunjucks-addons\(ra\fR package.
.SH "OPTIONS"
.P
The following top-level options and arguments exist:
.RS 0
.IP \(bu 4
\[lB]\fB-h\fR|\fB--help\fR\[rB] Show usage help.
.IP \(bu 4
\[lB]\fB-V\fR|\fB--version\fR\[rB] Show program version information.
.IP \(bu 4
\[lB]\fB-c\fR|\fB--config\fR \fIconfig-file\fR\[rB] Load Nunjucks configuration YAML file.
.IP \(bu 4
\[lB]\fB-C\fR|\fB--option\fR \fIkey\fR=\fIvalue\fR\[rB] Set Nunjucks configuration option.
.IP \(bu 4
\[lB]\fB-d\fR|\fB--defines\fR \fIcontext-file\fR\[rB] Load context definition YAML file. Can occur multiple times.
.IP \(bu 4
\[lB]\fB-D\fR|\fB--define\fR \fIkey\fR=\fIvalue\fR\[rB] Set context definition key/value. Can occur multiple times.
.IP \(bu 4
\[lB]\fB-e\fR|\fB--extension\fR \fImodule-name\fR\[rB] Load Nunjucks JavaScript extension module (installed via NPM).
.IP \(bu 4
\[lB]\fB-o\fR|\fB--output\fR \fIoutput-file\fR|\fB-\fR\[rB] Save output file (or stdout).
.IP \(bu 4
\[lB]\fB<input-file>\fR|\fB-\fR\[rB] Load input file (or stdin).
.RE 0

.SH "EXAMPLE"
.P
.RS 2
.nf
$ echo "Hello, {{who}}!" | nunjucks -D who=world -
Hello, world!
.fi
.RE
.SH "HISTORY"
.P
The \fBnunjucks\fR(1) utility was developed in August 2023 for being able to easily generate multiple configuration files for a complex \fIDocker-Compose\fR based setup.
.SH "AUTHOR"
.P
Dr. Ralf S. Engelschall \fI\(larse@engelschall.com\(ra\fR


nunjucks.js
#!/usr/bin/env node
/*
**  nunjucks -- Nunjucks Template Rendering Command-Line Interface
**  Copyright (c) 2019-2025 Dr. Ralf S. Engelschall <http://engelschall.com>
**  Licensed under MIT <http://spdx.org/licenses/MIT.html>
*/

/*  internal requirements  */
import fs          from "node:fs"
import path        from "node:path"

/*  external requirements  */
import { Command } from "commander"
import chalk       from "chalk"
import jsYAML      from "js-yaml"
import nunjucks    from "nunjucks"
import deepmerge   from "deepmerge"

/*  load my own information  */
const my = JSON.parse(await fs.promises.readFile(new URL("./package.json", import.meta.url)))

/*  parse command-line arguments  */
const program = new Command()
program.name("nunjucks")
    .description("Nunjucks Template Rendering Command-Line Interface")
    .showHelpAfterError("hint: use option --help for usage information")
    .option("-h, --help", "show usage help", false)
    .option("-V, --version", "show program version information", false)
    .option("-c, --config <config-file>", "load Nunjucks configuration YAML file", "")
    .option("-C, --option <key>=<value>", "set Nunjucks configuration option", (v, l) => l.concat([ v ]), [])
    .option("-d, --defines <context-file>", "load context definition YAML file", (v, l) => l.concat([ v ]), [])
    .option("-D, --define <key>=<value>", "set context definition key/value", (v, l) => l.concat([ v ]), [])
    .option("-e, --extension <module-name>", "load Nunjucks JavaScript extension module", (v, l) => l.concat([ v ]), [])
    .option("-o, --output <output-file>", "save output file", "-")
    .argument("[<input-file>]", "input file")
program.parse(process.argv)
const argv = { ...program.opts(), _: program.args }

/*  handle special help request  */
if (argv.help) {
    console.log(program.helpInformation())
    console.log("Example:\n  $ echo \"Hello, {{ who }}!\" | nunjucks -Dwho=World -\n")
    process.exit(0)
}

/*  handle special version request  */
if (argv.version) {
    console.log(`${my.name} ${my.version} (Node.js ${process.versions.node}, Nunjucks: ${my.dependencies.nunjucks})`)
    console.log(`${my.description}`)
    console.log(`Copyright (c) 2019-2025 ${my.author.name} <${my.author.url}>`)
    console.log(`Licensed under ${my.license} <http://spdx.org/licenses/${my.license}.html>`)
    process.exit(0)
}

/*  read input file  */
let input = ""
if (argv._.length > 1) {
    console.error(chalk.red("nunjucks: ERROR: invalid number of arguments (zero or one input file expected)"))
    process.exit(1)
}
let inputFile = argv._[0] ?? "-"
if (inputFile === "-") {
    inputFile = "<stdin>"
    process.stdin.setEncoding("utf-8")
    const BUFSIZE = 256
    const buf = Buffer.alloc(BUFSIZE)
    while (true) {
        let bytesRead = 0
        try {
            bytesRead = fs.readSync(process.stdin.fd, buf, 0, BUFSIZE)
        }
        catch (ex) {
            if      (ex.code === "EAGAIN") continue
            else if (ex.code === "EOF")    break
            else                           throw ex
        }
        if (bytesRead === 0)
            break
        input += buf.toString("utf8", 0, bytesRead)
    }
}
else {
    if (!fs.existsSync(inputFile)) {
        console.error(chalk.red(`nunjucks: ERROR: failed to find input file: ${inputFile}`))
        process.exit(1)
    }
    input = fs.readFileSync(inputFile, { encoding: "utf8" })
}

/*  provide context variables for template  */
let context = {}
for (const define of argv.defines) {
    try {
        context = deepmerge(context, jsYAML.load(fs.readFileSync(define, { encoding: "utf8" })))
    }
    catch (ex) {
        console.error(chalk.red(`nunjucks: ERROR: failed to load context YAML file: ${ex.toString()}`))
        process.exit(1)
    }
}

/*  expose environment variables to template  */
context.env = process.env

/*  add context defines  */
argv.define.forEach((define) => {
    let [ , key, val ] = define.match(/^([^=]+)(?:=(.*))?$/)
    if (val === undefined)
        val = true
    context[key] = val
})

/*  determine Nunjucks options  */
let options = {}
if (argv.config) {
    try {
        options = jsYAML.load(fs.readFileSync(argv.config, { encoding: "utf8" }))
    }
    catch (ex) {
        console.error(chalk.red(`nunjucks: ERROR: failed to load options YAML file: ${ex.toString()}`))
        process.exit(1)
    }
}
if (argv.option.length > 0)
    options = Object.assign(options, argv.option)
options = Object.assign({}, {
    autoescape:       false,
    throwOnUndefined: false,
    trimBlocks:       true,
    lstripBlocks:     true,
    watch:            false,
    noCache:          true
}, options)

/*  configure environment  */
const env = nunjucks.configure(inputFile, options)

/*  load external extension files  */
for (const extension of argv.extension) {
    let modpath = path.resolve(extension)
    if (!fs.existsSync(modpath)) {
        try {
            modpath = require.resolve(extension)
        }
        catch (ex) {
            modpath = null
        }
    }
    if (modpath === null) {
        console.error(chalk.red(`nunjucks: ERROR: failed to find extension module: ${extension}`))
        process.exit(1)
    }
    const mod = require(modpath)
    if (!(mod !== null && typeof mod === "function")) {
        console.error(chalk.red(`nunjucks: ERROR: failed to call extension file: ${modpath}`))
        process.exit(1)
    }
    mod(env)
}

/*  render Nunjucks template  */
let output
try {
    output = env.renderString(input, context)
}
catch (ex) {
    console.error(chalk.red(`nunjucks: ERROR: failed to render template: ${ex.toString()}`))
    process.exit(1)
}

/*  write output  */
if (argv.output === "-")
    process.stdout.write(output)
else
    fs.writeFileSync(argv.output, output, { encoding: "utf8" })

/*  die gracefully  */
process.exit(0)

